src/USER-MISC/fix_electron_stopping_fit.cpp

```diff
@@ -0,0 +1,202 @@
+   drag_fac_2 = new double[atom->ntypes+1];
+   
+   for (int i = 1; i <= atom->ntypes; i++) {
+      double mvv;
+      mvv = 2.0*energy_coh_in[i]/force->mvv2e;
+      
+      v_min_sq[i] = 1.0*mvv/atom->mass[i];
+      v_max_sq[i] = 2.0*mvv/atom->mass[i];
+      
+      drag_fac_1[i] = drag_fac_in_1[i];
+      drag_fac_2[i] = drag_fac_in_2[i];
+   };
+};
+
+// ---------------------------------------------------------------------
+
+FixElectronStoppingFit::~FixElectronStoppingFit()
+{
+   delete [] drag_fac_in_1;
+   delete [] drag_fac_in_2;
+   delete [] drag_fac_1;
+   delete [] drag_fac_2;
+   delete [] v_min_sq;
+   delete [] v_max_sq;
+};
+
+// ---------------------------------------------------------------------
+
+int FixElectronStoppingFit::setmask()
+{
+   int mask = 0;
+   mask |= POST_FORCE;
+   mask |= POST_FORCE_RESPA;
+   return mask;
+};
+
+// ---------------------------------------------------------------------
+
+void FixElectronStoppingFit::init()
+{
+   electronic_loss_this_node = 0.;
+   electronic_loss = 0.;
+   f_dot_v_prior = 0.;
+   f_dot_v_current = 0.;
+   last_step = update->ntimestep;
+};
+
+// ---------------------------------------------------------------------
+
+void FixElectronStoppingFit::setup(int vflag)
+{
+   if (strcmp(update->integrate_style,"verlet") == 0)
+      post_force(vflag);
+   else {
+      ((Respa *) update->integrate)->copy_flevel_f(nlevels_respa-1);
+      post_force_respa(vflag,nlevels_respa-1,0);
+      ((Respa *) update->integrate)->copy_f_flevel(nlevels_respa-1);
+   };
+};
+
+// ---------------------------------------------------------------------
+
+void FixElectronStoppingFit::post_force(int vflag)
+{
+   double **v = atom->v;
+   double **f = atom->f;
+   int *type  = atom->type;
+   int nlocal = atom->nlocal;
+   
+   f_dot_v_current = 0.0;
+   for (int i = 0; i < nlocal; i++) {
+      double vv = v[i][0]*v[i][0] + v[i][1]*v[i][1] + v[i][2]*v[i][2];
+      if (vv > v_min_sq[type[i]]) {
+         double gamma_x;
+         double gamma_y;
+         double gamma_z;
+         double v_mag = sqrt(vv);
+         if (vv < v_max_sq[type[i]]) {
+            double frac = (vv - v_min_sq[type[i]])/(v_max_sq[type[i]] - v_min_sq[type[i]]);
+            gamma_x = frac*(drag_fac_2[type[i]]*v[i][0] + drag_fac_1[type[i]]);
+       	    gamma_y = frac*(drag_fac_2[type[i]]*v[i][1] + drag_fac_1[type[i]]);
+       	    gamma_z = frac*(drag_fac_2[type[i]]*v[i][2] + drag_fac_1[type[i]]);
+         } else {
+            gamma_x = drag_fac_2[type[i]]*v[i][0] + drag_fac_1[type[i]];
+            gamma_y = drag_fac_2[type[i]]*v[i][1] + drag_fac_1[type[i]];
+            gamma_z = drag_fac_2[type[i]]*v[i][2] + drag_fac_1[type[i]];
+         };
+         f[i][0] -= gamma_x*v[i][0];
+         f[i][1] -= gamma_y*v[i][1];
+         f[i][2] -= gamma_z*v[i][2];
+         f_dot_v_current += sqrt(pow(gamma_x*v[i][0], 2.0) + pow(gamma_y*v[i][1], 2.0) + pow(gamma_z*v[i][2], 2.0))*v_mag;
```

Comment:
using the pow() function for computing a square is extremely slow. While a multiplication can be done with a single clock cycle, the pow() function requires computing a logarithm and an exponential and thus consumes hundreds of cycles and  on top of that produces a less precise result. You can use the MathSpecial::square() inline function in this case or MathSpecial::powint() for more general integer powers. This is where C/C++ differs from Fortran, which has the equivalent as built in (provided you use an integer and not a floating point number). Some modern C++ compilers provide overloaded std::pow() functions, but they are also dependent on using an integer power to access the (much) faster exponentiation. Yet for a simple square nothing can beat the square() inline function.

Reply:
pow() is now replaced with MathSpecial::square().